### PR TITLE
added: multi-volume support

### PR DESCRIPTION
**Pull request by notspiff**
- suppress everything but first file in multi-volume rar archives
- re-add rar support
- make the add-on also accept the rar:// protocol for backwards compatibility

this should be the nail in the coffin for vfs.rar \o/

### CODE REVIEW

File: src/ArchiveFile.cpp [hunk 1]
```diff
@@ -61,29 +62,90 @@ static std::string URLEncode(const std::string& strURLData)
...
-    bool Open()
+    bool Open(const std::string& url)
     {
-      if (!file.OpenFile(url))
-        return false;
+      // check for multi-volume rar
+      bool oldstyle = false;
+      if (url.find(".rar") != std::string::npos)
+      {
+        size_t partpos = url.find(".part");
```
Comment (pkerling) [inline src/ArchiveFile.cpp]: Since `partpos` is never used as position, you might just want to make it a `bool` to be clear

Comment (notspiff) [reply src/ArchiveFile.cpp]: had to rewrite to use re's so it's gone.

File: src/ArchiveFile.cpp [hunk 2]
```diff
@@ -61,29 +63,98 @@ static std::string URLEncode(const std::string& strURLData)
...
     {
-      if (!file.OpenFile(url))
-        return false;
+      // check for multi-volume rar
+      bool oldStyle = false;
+      std::regex ext_re(".+[/\\\\](.+\\.rar)$");
+      std::smatch match;
+      if (std::regex_match(url, match, ext_re))
+      {
+        std::regex part_re("(.+\\.part)[0-9]+\\.rar");
```
Comment (pkerling) [inline src/ArchiveFile.cpp]: Might want to add $ at the end for consistency?

Comment (notspiff) [reply src/ArchiveFile.cpp]: done

File: src/ArchiveFile.cpp [hunk 3]
```diff
@@ -61,29 +63,98 @@ static std::string URLEncode(const std::string& strURLData)
...
+            oldStyle = true;
+            fname = match[1].str();
+            fname.erase(fname.size()-3);
+          }
+        }
+
+        if (checkDir)
+        {
+          std::vector<kodi::vfs::CDirEntry> items;
+          size_t dirp = url.rfind('/');
```
Comment (pkerling) [inline src/ArchiveFile.cpp]: Wouldn't it be easier to also extract it with the regexp on L87?

File: src/ArchiveFile.cpp [hunk 4]
```diff
@@ -243,11 +332,25 @@ class CArchiveFile
...
+      std::string fname(url.filename);
+      size_t spos = fname.rfind('/');
+      if (spos == std::string::npos)
+        spos = fname.rfind('\\');
+      fname.erase(0, spos);
+      std::regex part_re("\\.part([0-9]+)\\.rar$");
+      std::smatch match;
+      if (std::regex_search(fname, match, part_re) && match.size() > 1)
+      {
+        if (atoi(match[1].str().c_str()) != 1)
```
Comment (pkerling) [inline src/ArchiveFile.cpp]: `std::stoul` takes `std::string` natively

File: src/ArchiveFile.cpp [hunk 5]
```diff
@@ -243,11 +332,25 @@ class CArchiveFile
...
+    if (strstr(url.filename, ".rar"))
+    {
+      std::string fname(url.filename);
+      size_t spos = fname.rfind('/');
+      if (spos == std::string::npos)
+        spos = fname.rfind('\\');
+      fname.erase(0, spos);
+      std::regex part_re("\\.part([0-9]+)\\.rar$");
+      std::smatch match;
+      if (std::regex_search(fname, match, part_re) && match.size() > 1)
```
Comment (pkerling) [inline src/ArchiveFile.cpp]: I don't think you need to check the size of `match`

File: src/ArchiveFile.cpp (hunk 3 above)
Comment (notspiff) [reply src/ArchiveFile.cpp]: done.

File: src/ArchiveFile.cpp (hunk 4 above)
Comment (notspiff) [reply src/ArchiveFile.cpp]: changed.

File: src/ArchiveFile.cpp (hunk 5 above)
Comment (notspiff) [reply src/ArchiveFile.cpp]: no you are correct, just old habit to check before deref.

**Pull request merged**